core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedStreamSourceChannel.java

```diff
@@ -263,35 +264,51 @@ public void addCloseTask(ChannelListener<R> channelListener) {
-                                    //we stop when writes are shutdown because we can't flush until we are active
-                                    //although we may be flushed as part of a batch
-                                    moreData = (frameDataRemaining > 0 && data != null) || !pendingFrameData.isEmpty() || anyAreSet(state, STATE_WAITNG_MINUS_ONE);
+            // mark state awaken if wakeup is true
+            if (wakeup)
+                state |= STATE_READS_AWAKEN;
+            // if not waked && not resumed, return
+            else if (!anyAreSet(state, STATE_READS_RESUMED))
+                return;
+            if (!anyAreSet(state, STATE_IN_LISTENER_LOOP)) {
+                state |= STATE_IN_LISTENER_LOOP;
+                getFramedChannel().runInIoThread(new Runnable() {
+
+                    @Override
+                    public void run() {
+                        try {
+                            boolean readAgain;
+                            do {
+                                synchronized(lock) {
+                                    state &= ~STATE_READS_AWAKEN;
+                                }
+                                ChannelListener<? super R> listener = getReadListener();
+                                if (listener == null || !isReadResumed()) {
+                                    return;
+                                }
+                                ChannelListeners.invokeChannelListener((R) AbstractFramedStreamSourceChannel.this, listener);
+                                //if writes are shutdown or we become active then we stop looping
+                                //we stop when writes are shutdown because we can't flush until we are active
+                                //although we may be flushed as part of a batch
+                                final boolean moreData = (frameDataRemaining > 0 && data != null) || !pendingFrameData.isEmpty() || anyAreSet(state, STATE_WAITNG_MINUS_ONE);
+
+                                synchronized (lock) {
+                                    // keep running if either reads are resumed and there is more data to read, or if reads are awaken
+                                    readAgain =((isReadResumed() && moreData) || allAreSet(state, STATE_READS_AWAKEN))
+                                               // as long as channel is not closed and there is no stream broken
+                                               && allAreClear(state,STATE_CLOSED | STATE_STREAM_BROKEN);
+                                    if (!readAgain)
+                                        state &= ~STATE_IN_LISTENER_LOOP;
                                 }
-                                while (allAreSet(state, STATE_READS_RESUMED) && allAreClear(state, STATE_CLOSED | STATE_STREAM_BROKEN) && moreData);
-                            } finally {
+                            } while (readAgain);
+                        } catch (RuntimeException e) {
```

Comment:
We might want to be cautious here and handle errors just in case.
```suggestion
                        } catch (RuntimeException | Error e) {
```

Reply:
+1, I'll incorporate this change